### ISSUE TITLE
CLE Lab: Add index selection in all library construction protocols

### ISSUE THREAD

**Issue by tim-snyder-wustl** [GTAC-MGI/GTAC-ESP-LIMS#210]
Waiting on more info from Andy.

- Need index integration in all library construction protocols - still thinking that index name might be best for picking but if we could add well location to name that would be nice
- Need to update Index DB with CLE barcodes (new column/property).

**Comment by abohanno** [comment]
I will try and work on adding index barcodes. If we could pick by barcode but then have it show us barcode, index name, and sequence that would be ideal

**Comment by LMT337** [comment]
Reached out to Andy to add barcod columns to index sheet: 
https://wustl.app.box.com/file/1009899994712
Asked for clarification on columns to pull in to ESP.

**Comment by LMT337** [comment]
Library Construction Workflows: 
CLE Lab: EpMotion Dilutions for Library Construction
No Construction Protocol

CLE Lab: Haloplex Library Construction
Protocol:
CLE Lab: Haloplex Library Construction

CLE Lab: Targeted Capture Library Construction
Protocol:
CLE Lab: KAPA Hyper Library Construction

CLE Lab: Whole Genome Library Construction
Protocol:
CLE Lab: ILMN Library Construction

**Comment by abohanno** [comment]
[CLE_IndexSequences.xlsx](https://github.com/GTAC-MGI/GTAC-ESP-LIMS/files/15485799/CLE_IndexSequences.xlsx)

**Comment by LMT337** [comment]
Need index ingest process, plate barcode, 2d barcodes to well position for master tray. 
Tray gets aliquoted into temp index tray. 
For index selection, there are three trays, DNA, index and destination. Plate scan those into a workflow sheet. 
Also way to manually seclect index if plate scanner is not available.

**Comment by LMT337** [comment]
WGS uses master tray, no aliquot tray.

**Comment by tim-snyder-wustl** [comment]
For Webtools:

- add option to 'decommision' plates - mark as 'empty' in the UI
- 2D barcodes will change, but indexes stay the same in each well
- exome master tray is static, gets used only once, the aliquot tray gets used ~18 times
- need to add 2D barcodes to each well in firestore

**Comment by billtell** [comment]
- [x] Moved Reagent tracking protocol
- [x] Moved Next Step to sample protocol

### **Basic Workflow Model**
- Plate Layout - VisionMate pipeline
> Confirm Dilution barcodes and position
> Set Library Barcodes based on Well Position
- Assign Barcodes - BQ lookup
> Query BQ for corresponding Index info
> Does Negative Control get index and reagents?
- Create Libraries
>1:1 creation; Set Library values

**Comment by billtell** [comment]
### Setup Protocols - Overall Reagent check

_**Kapa**_
Technicians as protocol field on for each sample
Microtube Plate
Pre and Post amplification -> Move to after library construction
> In other words,  First protocol - frag reagents , after Sample protcol - amplification reagents?

_**WG**_
ILMN Box 1, ILMN Box 2, ILMN Box 3
IDT 10bp Dual Indexes
pcr-free reagent add
- maybe have a dropdown that is required: pcr-free vs. amplified
- IDT 10bp Dual Indexes -> ILMN DNA UD Indexes A

**Comment by billtell** [comment]
using visionmate example file in #209 ... for now...🤞🏻

**Comment by billtell** [comment]
- [x] Basic VisionMate pipeline created

### Dilution VisionMate Check:
Usefullness?
Checks to make / Things to guard against

**Comment by billtell** [comment]
[visionmate-sample.csv](https://github.com/user-attachments/files/18399848/visionmate-sample.csv)
Example format for visionmate file

**Comment by tim-snyder-wustl** [comment]
- WGS plates should be automatically inactive when all the wells are used up
- Exome wells can be used multiple times, should increment a useage field or re-use the active field
- for exome plates - well ~18 rections per well 1728 total reactions per plate
- tech will upload updated visionmate file and pipeline script should map accordingly - tech will re-arrange samples in a plate and upload the new visionmate file.

**Comment by abohanno** [comment]
Hi All,

Not sure if this is where I need to put this info but that is what I have in my notes. Variables for ILMN Amplified libraries vs. ILMN PCR-Free libraries listed below:

![Image](https://github.com/user-attachments/assets/456c1f31-996d-4649-b62d-0dd3b81d588b)

![Image](https://github.com/user-attachments/assets/7f71ea4e-d49c-4e75-99ea-a327445cf3c9)

**Comment by billtell** [comment]
@abohanno , thanks.  this is the right spot.

**Comment by billtell** [comment]
@abohanno there is an agilent step in the amplified workflow?  is a file added?
agilent measures concentration, right?  will you need to bring in values from a file at this step?
or is that measure on the qubit, meaning we should reattach wgs to library assessment?

**Comment by abohanno** [comment]
@billtell the agilent only measures library size. Shelly mentioned in the meeting yesterday just adding a column to the protocol where we manually add the size. We would record the concentration data from a qubit reading. I would assume that we could just add this value as well instead of adding a whole node to the workflow chain. I doubt that this will be used often.

**Comment by billtell** [comment]
### CLE Lab: Whole Genome Library Construction Setup - Reagents to add

- [x] Qubit 1X Premix
- [x] Agilent Reagent Kit
- [x] Agilent Chip
- [x] DEPC Water

- [x] ILMN Box 3 is blank for PCR-Free

- [ ] Add Agilent File and Qubit File attachment at top of 1st or 3rd protocol
> Yes!

- For Amplified: manually enter size and concentration in sample protocol, otherwise leave blank.
> Isn't PCR-free `Size(bp)` a default value. 
> Set to 450!

- Use a `LIbrary Prep Type` picklist column, pcr-free default to pre-set a few reagent values and `Size (bp)`  
> made an onchange function  if value is changed back/forth from pcr-free & amplified - `onchange_wgs_reagent_presets`

**Comment by tim-snyder-wustl** [comment]
Re: Index Assignment:

- swap order of protocols, assign indexes *after* sample protocol
- Targeted Capture / Exome - map index barcodes to sample barcodes, assign indexes, update usage count in BQ
- WGS - just map index wells to sample wells, no barcodes needed. Indexes are always in the same wells. Update usage count in BQ.

**Comment by abohanno** [comment]
I made it through WGS library construction this afternoon. The assign indexes button worked!

In this protocol, do you know if the "Plate Barcode" and "Well Location" columns that follow the index sequence column are in reference to the indexes or the library? They are required currently and did not get filled in by the button. I wasn't sure what to put there.

**Comment by abohanno** [comment]
the web tools app was also updated to reflect the index wells that were used

**Comment by tim-snyder-wustl** [comment]
@abohanno those fields are in reference to the library. Are they correct? Do they need to be there?

**Comment by tim-snyder-wustl** [comment]
@abohanno 

Creating a separate ticket for the Targeted Capture index assignment work:

https://github.com/GTAC-MGI/GTAC-ESP-LIMS/issues/521

**Comment by abohanno** [comment]
I'll go back and review the WGS library construction protocols. We might not need those values there, or they could be auto-filled by a prior protocol.

**Comment by abohanno** [comment]
/opt/l7esp/.local/lib/python3.9/site-packages/google/cloud/bigquery/table.py:1820: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(
Traceback (most recent call last):
  File "/opt/l7esp/data/content/pipelines/scripts/cle_assign_indexes.py", line 109, in <module>
    job.result() 
  File "/opt/l7esp/.local/lib/python3.9/site-packages/google/cloud/bigquery/job/base.py", line 969, in result
    return super(_AsyncJob, self).result(timeout=timeout, **kwargs)
  File "/opt/l7esp/.local/lib/python3.9/site-packages/google/api_core/future/polling.py", line 261, in result
    raise self._exception
google.api_core.exceptions.BadRequest: 400 STRING(3) has maximum length 3 but got a value with length 4; in column 'well'; reason: invalidQuery, location: query, message: STRING(3) has maximum length 3 but got a value with length 4; in column 'well'
{
 "units": "ms"
 "approximate_execution_time": 35526.660216
 "call_headers": ['API', '#Calls', 'Cumulative', 'Min', 'Max', 'Mean', '#UniqueEndpoints']
 "calls": [
  ["PUT /api/sample_sheets/:uuid/set_sample_values", 1, 20798.522716, 20798.522716, 20798.522716, 20798.522716, 1]
  ["GET /api/sample_sheets/:uuid/bulk", 2, 667.101165, 285.704627, 381.396538, 333.5505825, 1]
  ["GET /api/workflow_instances/:uuid", 1, 564.968198, 564.968198, 564.968198, 564.968198, 1]
  ["GET /api/sample_sheets/:uuid", 2, 380.397932, 186.546556, 193.851376, 190.198966, 1]
  ["GET /api/projects/:uuid", 1, 205.783768, 205.783768, 205.783768, 205.783768, 1]
  ["GET /api/samples", 1, 122.354461, 122.354461, 122.354461, 122.354461, 1]
 ]
}

received above error at assign indexes step in protocol 3 of WGS library construction

**Comment by tim-snyder-wustl** [comment]
@abohanno fixed and deployed to Sandbox.
